data.go

```diff
@@ -71,7 +73,86 @@ func (r *dataReader) Read(b []byte) (n int, err error) {
 		}
 	}
 
-	n, err = r.r.Read(b)
+	// Code below is taken from net/textproto with only one modification to
+	// not rewrite CRLF -> LF.
+
+	// Run data through a simple state machine to
+	// elide leading dots and detect ending .\r\n line.
+	const (
+		stateBeginLine = iota // beginning of line; initial state; must be zero
+		stateDot              // read . at beginning of line
+		stateDotCR            // read .\r at beginning of line
+		stateCR               // read \r (possibly at end of line)
+		stateData             // reading data in middle of line
+		stateEOF              // reached .\r\n end marker line
+	)
+	for n < len(b) && r.state != stateEOF {
+		var c byte
+		c, err = r.r.ReadByte()
+		if err != nil {
+			if err == io.EOF {
+				err = io.ErrUnexpectedEOF
+			}
+			break
+		}
+		switch r.state {
+		case stateBeginLine:
+			if c == '.' {
+				r.state = stateDot
+				continue
+			}
+			if c == '\r' {
+				r.state = stateCR
+				continue
+			}
+			r.state = stateData
+
+		case stateDot:
+			if c == '\r' {
+				r.state = stateDotCR
+				continue
+			}
+			if c == '\n' {
+				r.state = stateEOF
+				continue
+			}
+			r.state = stateData
+
+		case stateDotCR:
+			if c == '\n' {
+				r.state = stateEOF
+				continue
+			}
+			// Not part of .\r\n.
+			// Consume leading dot and emit saved \r.
+			r.r.UnreadByte()
```

Comment:
We don't "save" the `\r` anymore, since we don't skip appending to `b` here: https://github.com/emersion/go-smtp/pull/111/files#diff-73850c2e1fb203a205030b007f382b14R144

So, I believe this trips on `.\rasdf`, and emits a duplicate `\r`. I think we can fix it by just updating state to `stateData` without calling `UnreadByte` and without changing the value of `c`.